File: Comparative/testing/goImp/above60/problem384_Rudin_Shapiro_Sequence.go
// https://projecteuler.net/problem=384
// (65%)

package main

import (
	"fmt"
)

func count(n, x int64) int64 {
	ret := int64(0)
	start := int64(0)
	now := int64(0)

	for hi := 62; hi >= 0; hi-- {
		if (n >> hi) & 1 != 0 {
			if now & 1 != 0 {
				start = -start
				x = -x
			}

			mid := start
			dff := int64(1 << ((hi + 1) >> 1))

			if hi & 1 != 0 {
				d := x - start
				if d < dff {
					ret += max(d, 0)
				} else if d == dff {
					ret += dff / 2
				}
				start += dff
			} else {
				mid += dff
				ret += max(0, dff - abs(mid - x))
				start = mid
			}

			if now & 1 != 0 {
				start = -start
				x = -x
			}

			if (n >> hi) % 4 == 3 {
				now++
			}
		}
	}

	return ret
}

func pos(x, y int64) int64 {
	L, R := int64(1), int64(1 << 62 - 1)
	for L < R {
		m := (L + R) / 2
		if count(m, x) >= y {
			R = m
		} else {
			L = m + 1
		}
	}
	return L
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func abs(a int64) int64 {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	a, b := int64(1), int64(2)
	ans := int64(0)

	for i := 2; i < 46; i++ {
		t := pos(b, a)
		// fmt.Println(t, count(t, b), a)
		ans += t - 1
		c := b + a
		a, b = b, c
	}

	fmt.Println(ans)
	// fmt.Println(count(1220847711, 54321), pos(54321, 12345) - 1)
}

File: Comparative/testing/kotlinImp/above60/problem384_Rudin_Shapiro_Sequence.kt
// https://projecteuler.net/problem=384
// (65%)

import kotlin.math.*

class Counter {
    fun count(n: Long, x: Int): Long {
        var ret: Long = 0
        var start: Long = 0
        var now: Long = 0
        var xNew = x

        for (hi in 62 downTo -1) {
            if ((n shr hi) and 1L != 0L) {
                if (now and 1L != 0L) {
                    start = -start
                    xNew = -xNew
                }

                var mid = start
                val dff: Long = 1L shl ((hi + 1) shr 1)

                if (hi and 1 != 0) {
                    val d = xNew - start
                    if (d < dff) {
                        ret += maxOf(d, 0)
                    } else if (d == dff) {
                        ret += dff / 2
                    }
                    start += dff
                } else {
                    mid += dff
                    ret += maxOf(0, dff - abs(mid - xNew))
                    start = mid
                }

                if (now and 1L != 0L) {
                    start = -start
                    xNew = -xNew
                }

                now += if ((n shr hi.toInt()) % 4L == 3L) 1 else 0
            }
        }

        return ret
    }

    fun pos(x: Int, y: Int): Long {
        var L: Long = 1
        var R: Long = (1L shl 62) - 1
        while (L < R) {
            val m = (L + R) / 2
            if (count(m, x) >= y) {
                R = m
            } else {
                L = m + 1
            }
        }
        return L
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val counter = Counter()
            var a = 1
            var b = 2
            var ans: Long = 0

            for (i in 2 until 46) {
                val t = counter.pos(b, a)
                ans += t - 1
                val c = b + a
                a = b
                b = c
            }

            println(ans)
        }
    }
}
